File: js/lib_bejson_gaming/lib_bejson_tile_grids.js
/**
 * switch_grid.js
 * game1 aka the BEJSON Game Engine (#1)
 * Author: Elton Boehnen: BEJSON-Native Tile Grid Manager (v1.0)
 * Date: 2026-05-03
 */

window.Switch = window.Switch || {};

class SwitchGrid {
    constructor(name, width, height) {
        this.width = width;
        this.height = height;
        this.bejson = Switch.BEJSON.create104(name, [
            { name: "layer_name", type: "string" },
            { name: "data", type: "array" }
        ], []);
    }

    createLayer(name, initialValue = 0) {
        const data = new Array(this.width * this.height).fill(initialValue);
        this.bejson.Values.push([name, data]);
    }

    getTile(layerName, x, y) {
        const layer = this.bejson.Values.find(v => v[0] === layerName);
        if (!layer || x < 0 || x >= this.width || y < 0 || y >= this.height) return null;
        return layer[1][y * this.width + x];
    }

    setTile(layerName, x, y, val) {
        const layer = this.bejson.Values.find(v => v[0] === layerName);
        if (!layer || x < 0 || x >= this.width || y < 0 || y >= this.height) return;
        layer[1][y * this.width + x] = val;
    }
}

Switch.Grid = SwitchGrid;

File: lib/py/Gaming/lib_bejson_tile_grids.py
"""
game1 aka the BEJSON Game Engine (#1)
Author: Elton Boehnen
"""
class BEJSONTileGrid:
    """
    Mirror of lib_bejson_tile_grids.js
    Python/Flask-compatible tile grid manager using BEJSON 104.
    """
    def __init__(self, grid_name="UntitledGrid", width=0, height=0):
        self.width = width
        self.height = height
        self.bejson = {
            "Format": "BEJSON",
            "Format_Version": "104",
            "Schema_Name": grid_name,
            "Records_Type": ["Layer"],
            "Fields": [
                { "name": "layer_name", "type": "string" },
                { "name": "data", "type": "array" }
            ],
            "Values": []
        }

    def create_layer(self, name, initial_value=0):
        data = [initial_value] * (self.width * self.height)
        self.bejson["Values"].append([name, data])

    def get_tile(self, layer_name, x, y):
        layer = next((row for row in self.bejson["Values"] if row[0] == layer_name), None)
        if not layer or x < 0 or x >= self.width or y < 0 or y >= self.height:
            return None
        return layer[1][y * self.width + x]

    def set_tile(self, layer_name, x, y, value):
        layer = next((row for row in self.bejson["Values"] if row[0] == layer_name), None)
        if not layer or x < 0 or x >= self.width or y < 0 or y >= self.height:
            return
        layer[1][y * self.width + x] = value

    def export_bejson(self):
        import json
        return json.dumps(self.bejson, indent=2)
